lib/types.ts

```diff
@@ -1,40 +1,47 @@
-import type {
-  TypeFromJSONSchema,
-  JSONSchema7,
-  JSONSchema7Definition,
-} from '@profusion/json-schema-to-typescript-definitions';
+import type { TypeFromJSONSchema } from '@profusion/json-schema-to-typescript-definitions';
+import type { JSONSchema7, JSONSchema7Definition } from 'json-schema';
+import type { DeepReadonly } from 'json-schema-to-ts/lib/types/type-utils/readonly';
+
+type ReadonlyJSONSchema = DeepReadonly<JSONSchema7>;
+type ReadonlyJSONSchemaDefinition = DeepReadonly<JSONSchema7Definition>;
+
+// string-only keys
+export type KeyOf<T extends object> = keyof T & string;
 
 export type BaseEnvSchema = {
   type: 'object';
-  properties: Readonly<{ [key: string]: JSONSchema7 }>;
+  properties: Readonly<{ [key: string]: ReadonlyJSONSchema }>;
   required?: readonly string[];
-  dependencies?: {
-    readonly [key: string]: JSONSchema7Definition | readonly string[];
-  };
+  dependencies?: Readonly<{
+    [key: string]: ReadonlyJSONSchemaDefinition | string[];
+  }>;
   additionalProperties?: true; // if provided, must be true, otherwise it may hurt process.env
 };
 
-export type EnvSchemaProperties<S extends BaseEnvSchema> = {
-  [K in Extract<keyof S['properties'], string>]: [K, S['properties'][K]];
-}[Extract<keyof S['properties'], string>][];
+export type BaseEnvParsed<S extends BaseEnvSchema> = {
+  [key: string]: unknown;
+} & Record<KeyOf<S['properties']>, unknown>;
 
-export const schemaProperties = <S extends BaseEnvSchema>(
-  schema: S,
-): Readonly<EnvSchemaProperties<S>> =>
-  Object.entries(schema.properties) as EnvSchemaProperties<S>;
+export type EnvSchemaProperties<
+  S extends BaseEnvSchema,
+  V extends BaseEnvParsed<S> = TypeFromJSONSchema<S>,
+  Keys extends KeyOf<S['properties'] & V> = KeyOf<S['properties'] & V>,
+> = {
+  [K in Keys]: [K, S['properties'][K]];
+}[Keys][];
 
-/**
- * Subset of valid (post-validate) properties. If there are no
- * errors, then all required properties should be present.
- */
-export type EnvSchemaPartialValues<S extends BaseEnvSchema> = Partial<
-  TypeFromJSONSchema<S>
->;
+export const schemaProperties = <
+  S extends BaseEnvSchema,
+  V extends BaseEnvParsed<S> = TypeFromJSONSchema<S>,
+  Ret = EnvSchemaProperties<S, V>,
+>({
+  properties,
+}: S): Readonly<Ret> => Object.entries(properties) as Ret;
 
 export type EnvSchemaPropertyValue<
   S extends BaseEnvSchema,
-  K extends keyof S['properties'],
-> = K extends keyof TypeFromJSONSchema<S> ? TypeFromJSONSchema<S>[K] : never;
+  V extends BaseEnvParsed<S>,
+> = V[KeyOf<V>];
```

Comment:
Ultimately, you would simply do
```ts
export type EnvSchemaPropertyValue<
  S extends BaseEnvSchema,
> = TypeFromJSONSchema<S>[KeyOf<TypeFromJSONSchema<S>>];
```
If that's possible, or makes sense. Well, if in fact I miss something and there is some case where you don't use the default value `=TypeFromJSONSchema<S>`, then this is necessary, but I haven't found

Reply:
What I'm trying to avoid here is the previous behavior of repeatedly calling `TypeFromJSONSchema<>`, processing the same conversion over and over. Which is what made this codebase so slow to work with..